doc/man7/flux-jobtap-plugins.rst

```diff
@@ -0,0 +1,137 @@
+
+   int callback (flux_plugin_t *p,
+                 const char *topic,
+                 flux_plugin_arg_t *args,
+                 void *arg);
+
+where ``p`` is the handle for the current *jobtap* plugin, ``topic`` is
+the *topic string* for the currently invoked callback, ``args`` contains
+a set of plugin arguments which may be unpacked with the
+``flux_plugin_arg_unpack(3)`` call, and ``arg`` is any opaque argument
+passed along when registering the handler.
+
+JOBTAP PLUGIN ARGUMENTS
+=======================
+
+For job-specific callbacks, all job data is passed to the plugin via
+the ``flux_plugin_arg_t *args``, and return data is sent back to the
+job manager via the same ``args``. Incoming arguments may be unpacked
+using ``flux_plugin_arg_unpack(3)``, e.g.::
+
+   rc = flux_plugin_arg_unpack (args, FLUX_PLUGIN_ARG_IN,
+                                "{s{s:o}, s:I}",
+                                "jobspec", "resources", &resources,
+                                "id", &id);
+
+will unpack the ``resources`` section of jobspec and the jobid into
+``resources`` and ``id`` respectively.
+
+The full list of available args includes the following:
+
+========== ==== ==========================================
+name       type description
+========== ==== ==========================================
+jobspec    o    jobspec with environment redacted
+id         I    jobid
+state      i    current job state
+prev_state i    previous state (``job.state.*`` callbacks)
+userid     i    userid
+urgency    i    current urgency
+priority   I    current priority
+t_submit   f    submit timestamp in floating point seconds
+entry      o    posted eventlog entry, including context
+========== ==== ==========================================
+
+Return arguments can be packed using the ``FLUX_PLUGIN_ARG_OUT`` and
+optionally ``FLUX_PLUGIN_ARG_UPDATE`` flags. For example to return
+a priority::
+
+   rc = flux_plugin_arg_pack (args, FLUX_PLUGIN_ARG_OUT,
+                              "{s:I}",
+                              "priority", (int64_t) priority);
+
+While a job is pending, *jobtap* plugin callbacks may also add job
+annotations by returning a value for the ``annotations`` key::
+
+   flux_plugin_arg_pack (args, FLUX_PLUGIN_ARG_OUT|FLUX_PLUGIN_ARG_UPDATE,
+                         "{s:{s:s}}",
+                         "annotations", "test", value);
+
+.. _callback_topics:
+
+CALLBACK TOPICS
+===============
+
+The following callback "topic strings" are currently provided by the
+*jobtap* interface:
+
+job.new
+  The ``job.new`` topic is called whenever the job manager is adding a new
+  job to its internal state. This callback is called in two scenarios: on
+  job submission, in which case the job state will be ``FLUX_JOB_STATE_NEW``,
+  or when the job manager is restarting and has reloaded a job from the KVS,
+  in which case the job state could be any state except
+  ``FLUX_JOB_STATE_INACTIVE``.
```

Comment:
assuming `job.new` cannot annotate, should note that here?  or above in annotation notes.